[sql-33] session: add migration code from kvdb to SQL

ViktorTigerstrom opened the pull request:
This PR introduces the migration logic for transitioning the sessions store from kvdb to SQL.

Note that as of this PR, the migration is not yet triggered by any production code, i.e. only tests execute the migration logic.

Part of https://github.com/lightninglabs/lightning-terminal/issues/917

ViktorTigerstrom commented on session/sql_store.go:
```diff
@@ -745,9 +745,12 @@ func unmarshalSession(ctx context.Context, db SQLQueries,
-		macRecipe = &MacaroonRecipe{
-			Permissions: unmarshalMacPerms(perms),
-			Caveats:     unmarshalMacCaveats(caveats),
+		macRecipe = &MacaroonRecipe{}
+		if perms != nil {
+			macRecipe.Permissions = unmarshalMacPerms(perms)
+		}
+		if caveats != nil {
+			macRecipe.Caveats = unmarshalMacCaveats(caveats)
```
Let me know if you think we should rather fix the kvdb store implementation rather than the sql one. But as we've previously opted for not updating the store behaviour, and make the sql one mimic the behaviour until we've completed the migration, I opted for this version here.

ViktorTigerstrom replied:
Alternatively, I could also override this in newly added `overrideMacaroonRecipe` function before the migration deep equals check. But I kept this as is, as I do think it's correct behaviour that we don't set the `Permissions` or `Caveats` and keep them at `nil`, if no value actually exists for them.

ViktorTigerstrom replied:
Please see https://github.com/lightninglabs/lightning-terminal/pull/1051#issuecomment-2934569218

ViktorTigerstrom commented on session/sql_migration.go:
```diff
@@ -0,0 +1,403 @@
+	// Since the InsertSession query doesn't support that we set the revoked
+	// field during the insert, we need to set the field after the session
+	// has been created.
+	if !session.RevokedAt.IsZero() {
+		err = tx.SetSessionRevokedAt(
+			ctx,
+			makeSetRevokedAtParams(sqlId, session.RevokedAt),
+		)
+		if err != nil {
+			return err
+		}
+	}
```
Could potentially update the insert session sql query to also take an optional `RevokedAt sql.NullTime` param, to not have to do this as a separate step. I opted for this version to keep things as is as much as possible, but if you think that'd be better, I'll update to do that instead :)

ViktorTigerstrom commented on session/sql_migration_test.go:
```diff
@@ -0,0 +1,738 @@
+				require.NoError(t, err)
+			},
+		},
+		{
+			"randomized sessions",
+			randomizedSessions,
+		},
```
Opted for a test that uses the "standard" `rand` lib here, as @bitromortac had a preference for that in https://github.com/lightninglabs/lightning-terminal/pull/1047#discussion_r2059882334. But can also look into adding a test that uses `rapid` if you'd prefer.

ViktorTigerstrom commented on session/sql_migration_test.go:
```diff
@@ -0,0 +1,738 @@
+		for i := 0; i < loopLen; i++ {
+			var macCaveat macaroon.Caveat
+
+			// We always have a caveat.Id, but the rest are
+			// randomized if they exist or not.
+			macCaveat.Id = randomBytes(rand.Intn(10) + 1)
```
Let me know if this **is not true** in for real world data, as migrations WILL fail if we don't set a `caveat.Id` when the `caveats` param of `WithMacaroonRecipe` is not `nil`.

ViktorTigerstrom replied:
Could potentially override this as well in the new `overrideMacaroonRecipe` as well. But I do not think this scenario can happen for real world data. But open to implementing that as well if you think it's worth doing it just to be sure.

ViktorTigerstrom commented on session/sql_migration_test.go:
```diff
@@ -0,0 +1,738 @@
+// "dest" state, without checking if the state transition is legal.
+//
+// NOTE: this function should only be used for testing purposes.
+func (db *BoltStore) shiftStateUnsafe(_ context.Context, id ID,
```
Not fully sure if we add test functions like this one that's a `struct` member to the `_test` file, or if we add it in the actual struct file?

bitromortac replied:
right, not too great, you could change to `func shiftStateUnsafe(_ context.Context, db *BoltStore, id ID, dest State) error {`

ViktorTigerstrom reviewed (commented):
Adding some review comments where input would be extra helpful :)!

bitromortac reviewed (commented):
Great work, very clean on first pass! :fire:

ellemouton reviewed (commented):
holding off on review here until the prep PR discussed in the accounts mig PR is in (since that will be used to test this one)

lightninglabs-deploy commented:
@viktortigerstrom, remember to re-request review from reviewers when ready

ViktorTigerstrom commented:
Now that #1047 has been merged, I'll update this PR to adhere to the feedback brought up in that PR, and then re-request reviews afterwards.

lightninglabs-deploy commented:
@viktortigerstrom, remember to re-request review from reviewers when ready

ViktorTigerstrom commented:
Updated the PR to also address feedback raised in #1047! Will address the small linting issue after next review round :)

bitromortac commented on session/sql_migration_test.go:
```diff
@@ -0,0 +1,368 @@
+)
+
+// TestSessionsStoreMigration tests the migration of session store from a bolt
+// backed to a SQL database. Note that this test does not attempt to be a
```
nit: `backend`

bitromortac reviewed (approved):
LGTM, utACK :tada:

ViktorTigerstrom commented on session/sql_migration_test.go:
```diff
@@ -0,0 +1,776 @@
+			},
+		},
+		{
+			name: "macaroon recipe with nil perms and caveats",
```
Note that this test is new, as this can actually happen for real world sessions.

ViktorTigerstrom reviewed (commented):
Thanks a lot for the review @bitromortac :tada:!! I noticed that we can have real world data that sets a non-nil macaroon recipe, but with `nil` `perms`& `caveats`. I updated the code to be able to handle that case as well.

ellemouton commented on session/sql_migration.go:
```diff
@@ -0,0 +1,364 @@
+		return err
+	}
+
+	// If sessions are linked to a group, we must insert the initial session
+	// of each group before the other sessions in that group. This ensures
+	// we can retrieve the SQL group ID when inserting the remaining
+	// sessions. Therefore, we first insert all initial group sessions,
```
ie, i think we can just:

```
for s := sessions {
       if s.groupID  = s.ID {
                    <insert new session. no coupled group>
                   continue
       }
       
       groupID := db.GetGroupIDByAlias(s.groupID)
       
        <insert new session. couple to group>
}  
```

ellemouton replied:
or did you actually run into a scenario where these were not in order?

ellemouton replied:
im also more in favour of doing it this way cause then the IDs that the sessions have will be more intuitive as then (as long as the user didnt do anything strange) , the created at times will mostly be increasing at the IDs increase

ViktorTigerstrom commented on session/sql_migration.go:
```diff
@@ -0,0 +1,364 @@
+			return
+		}
+
+		acctID = sql.NullInt64{
```
Thanks, addressed these and others :pray:!

ViktorTigerstrom replied:
Thanks, addressed these and others :pray:!

ViktorTigerstrom commented on session/sql_migration.go:
```diff
@@ -0,0 +1,364 @@
+		return err
+	}
+
+	// The remote public key is currently only set for autopilot sessions,
```
Ah yeah of course 😅... My bad, thanks :pray:!

ViktorTigerstrom commented on session/sql_migration.go:
```diff
@@ -0,0 +1,364 @@
+
+		err = tx.InsertSessionPrivacyFlag(ctx, privacyFlagParams)
+		if err != nil {
+			return err
```
Addressed this and others :)

ViktorTigerstrom commented:
I added a commit 6b8c58f9c9168471714adbba78446047b8614137 at the end of the PR to address the feedback of https://github.com/lightninglabs/lightning-terminal/pull/1051#discussion_r2111840035, as well as a fixup commit I will apply if you think the approach of added commit is good @ellemouton. Would therefore appreciate feedback on if this approach is what you had in mind when you suggested that I should change the kvdb implementation? Note that this actually changes how the `MacaroonRecipe` is persisted in `kvdb`, and it will set the field to `null` (which will also be represented when listing the session with `litcli`). 

Note that I will fix the failing unit test tomorrow, that's due to my update of how kvdb implementation stores the `MacaroonRecipe`.

ViktorTigerstrom commented:
Updated the PR by ensuring that the `overrideMacaroonRecipe` by itself handles the differences in how the `MacaroonRecipe` is represented between the kvdb & the SQL store. Therefore no changes are needed for how we'd persist it for neither the `kvdb` nor the `sql` stores.

@ellemouton, please let me know if you prefer that approach, or the other approach I've also implemented that changes how the kvdb persists the `MacaroonRecipe`. 
Personally I'm in strong favour of letting the `overrideMacaroonRecipe` func handle the differences, and i.e. drop the last 2 commits that currently exists in this PR.

ellemouton commented on session/sql_migration.go:
```diff
@@ -0,0 +1,361 @@
+
+		// Validate that the session was correctly migrated and matches
+		// the original session in the kv store.
+		sqlSess, err := tx.GetSessionByAlias(ctx, kvSession.ID[:])
+		if err != nil {
+			if errors.Is(err, sql.ErrNoRows) {
+				err = ErrSessionNotFound
+			}
+			return fmt.Errorf("unable to get migrated session "+
+				"from sql store: %w", err)
+		}
+
+		migratedSession, err := unmarshalSession(ctx, tx, sqlSess)
+		if err != nil {
+			return fmt.Errorf("unable to unmarshal migrated "+
+				"session: %w", err)
+		}
+
```
perhaps it was on the other PR that i commented this but i think we should first complete the whole migration before doing the validation no?

ellemouton replied:
ah, i see invoices was done like this too... ok all good

ellemouton reviewed (commented):
shweet! a few more comments. 

Agreed with you that the override macaroon recipe approach is good 👍  i think i didnt understand correctly before that for bbolt we could still have the TLV actually be written with out any contents (right?) so yes, the overwrite makes sense rather than the "un-set"

ellemouton commented on session/interface.go:
```diff
@@ -329,5 +329,8 @@ type Store interface {
 	// "dest" state.
 	ShiftState(ctx context.Context, id ID, dest State) error
 
+	// Close closes the underlying store.
+	Close() error
+
```
ah, i think you're doing it cause of test cleanup - rather just call t.Cleanup sooner?

ViktorTigerstrom commented:
Also, re-requested a review from @bitromortac, as quite much has changed since your last review. Hope that's ok!

ellemouton reviewed (approved):
lgtm! 
great work

bitromortac commented on session/sql_migration.go:
```diff
@@ -0,0 +1,396 @@
+		kvCaveats := kvSession.MacaroonRecipe.Caveats
+
+		if kvPerms == nil && kvCaveats == nil {
+			migratedSession.MacaroonRecipe = &MacaroonRecipe{}
```
In `GetSession` we use
```
	var macRecipe *MacaroonRecipe
	if perms != nil || caveats != nil {
		macRecipe = &MacaroonRecipe{
			Permissions: unmarshalMacPerms(perms),
			Caveats:     unmarshalMacCaveats(caveats),
		}
	}
```
but couldn't we just leave away the nil checks and `unmarshalX` nevertheless? this at least would give a bit more consistency (besides the empty slice vs nil slice behavior)

bitromortac commented on session/sql_migration_test.go:
```diff
@@ -379,3 +387,390 @@ func TestSessionsStoreMigration(t *testing.T) {
+		numberOfSessions = 100
+	)
+
+	for i := 0; i < numberOfSessions; i++ {
```
nit: there are a few occurrences here and afterwards where the new loop syntax could be used: `	for i := range numberOfSessions {`

ViktorTigerstrom replied:
Thanks addressed :)!

bitromortac reviewed (approved):
Nice, LGTM :tada:, just a question and a nit.

ViktorTigerstrom commented on accounts/sql_migration_test.go:
```diff
@@ -39,9 +39,6 @@ func TestAccountStoreMigration(t *testing.T) {
 		*db.TransactionExecutor[SQLQueries]) {
```
Did it that way so this wouldn't get merged without closing the stores in the tests, but I agree that it wouldn't be strictly necessary for merging :)

ViktorTigerstrom commented:
Thanks for the reviews :tada::zap:!
